pwndbg/heap/__init__.py

```diff
@@ -1,12 +1,81 @@
+import pwndbg.color.message as message
+import pwndbg.config
 import pwndbg.heap.heap
 import pwndbg.symbol
 
 current = None
 
 heap_chain_limit = pwndbg.config.Parameter('heap-dereference-limit', 8, 'number of bins to dereference')
 
+resolve_via_heuristic = pwndbg.config.Parameter('resolve-via-heuristic', False,
+                                                'Resolve some missing symbols via heuristics')
+
+main_arena = pwndbg.config.Parameter('main_arena', "0", 'main_arena address for heuristics')
+
+thread_arena = pwndbg.config.Parameter('thread_arena', "0", 'thread_arena value for heuristics')
+
+mp_ = pwndbg.config.Parameter('mp_', "0", 'mp_ address for heuristics')
+
+tcache = pwndbg.config.Parameter('tcache', "0", 'tcache value for heuristics')
+
+global_max_fast = pwndbg.config.Parameter('global_max_fast', "0", 'global_max_fast address for heuristics')
+
+symbol_list = [main_arena, thread_arena, mp_, tcache, global_max_fast]
+
+
+@pwndbg.config.Trigger(symbol_list)
+def parse_config2address():
+    # Somehow, we can't use integer config because the integer is too big and will cause out of range error for GDB API
+    # So we convert the string to a int manually
+    for symbol in symbol_list:
+        if not isinstance(symbol.value, str):
+            continue
+        address_str = symbol.value.strip()
+        if address_str == "0":
+            continue
+        try:
+            if address_str.startswith("0b"):
+                address = int(address_str, 2)
+            elif address_str.startswith("0o"):
+                address = int(address_str, 8)
+            elif address_str.startswith("0x"):
+                address = int(address_str, 16)
+            else:
+                address = int(address_str)
+        except ValueError:
+            symbol.value = "0"
+            raise ValueError("Please input a valid integer literal string")
+        symbol.value = address
+
+
 @pwndbg.events.start
 def update():
+    resolve_heap(is_first_run=True)
+
+
+@pwndbg.events.stop
+@pwndbg.events.new_objfile
+def reset():
```

Comment:
What's the reason to reset on each stop?

Reply:
That's because I want to make sure we refresh the heap like we update `pwndbg.typeinfo` when the file changes or process stopped, so we can avoid using some old addresses or wrong structs definitions.(btw my commit about it is https://github.com/pwndbg/pwndbg/commit/d89e484ba0e47b44fad470a5e29dbc9add2113d2)
But a little side effect is that not only when the file we want to debug changed, when loading some shared object files, the heap will keep refreshing too, and this wastes some resources.
So I think this part might need a better implementation, do you have any suggestions about this, or should I keep this part?
BTW, the way I use `importlib.reload` to reload structs seems a sort of dirty to me, should I write all the variables in an `update()` like what we did in `pwndbg.typeinfo`, it seems a little bit cleaner to me, but I'm not sure:
https://github.com/pwndbg/pwndbg/blob/e867d1db0c51017d5df927d86c33dd6b9b733456/pwndbg/typeinfo.py#L42-L47

Reply:
RE: better idea for this - not on top of my head, I will try to see if we have anything like that elsewhere.
\
Another idea could be only to mark that heap is "dirty" here and refresh it only on demand when someone asks for information regarding it. I guess that would be better?

Reply:
Regarding `importlib.reload` - yeah, I also think it is not the best approach, but we also do it in one of `update`, to reset elf types or something like that, if I recall correctly.

(I'm totally up for not using `reload` at all in those cases, so if u want to move it to `update` I'm up for that)

Reply:
> refresh it only on demand when someone asks for information regarding it. I guess that would be better?

Yes, I agree!
But sadly, after a few days of testing, I didn't figure out a good way to do it 😢
I'll update this part if I have a better ideas in the future.

> Regarding `importlib.reload` - yeah, I also think it is not the best approach, but we also do it in one of `update`, to reset elf types or something like that, if I recall correctly.
> 
> (I'm totally up for not using `reload` at all in those cases, so if u want to move it to `update` I agree on that)

Ok, I see.
And after some testing, I found that using `update()` might cause some errors in this case actually, so I'll keep the part of `importlib.reload`!